components/script/dom/intersectionobserver.rs

```diff
@@ -271,6 +302,352 @@ impl IntersectionObserver {
+        let intersection_rect = rect_to_domrectreadonly(intersection_rect);
+
+        // Step 1-2
+        // > 1. Construct an IntersectionObserverEntry, passing in time, rootBounds,
+        // >    boundingClientRect, intersectionRect, isIntersecting, and target.
+        // > 2. Append it to observer’s internal [[QueuedEntries]] slot.
+        self.queued_entries.borrow_mut().push(
+            IntersectionObserverEntry::new(
+                self.owner_doc.window(),
+                None,
+                document
+                    .owner_global()
+                    .performance()
+                    .to_dom_high_res_time_stamp(time),
+                Some(&root_bounds),
+                &bounding_client_rect,
+                &intersection_rect,
+                is_intersecting,
+                is_visible,
+                Finite::wrap(intersection_ratio),
+                target,
+                can_gc,
+            )
+            .as_traced(),
+        );
+        // > Step 3
+        // Queue an intersection observer task for document.
+        document.queue_an_intersection_observer_task();
+    }
+
+    /// Step 3.1-3.5 of <https://w3c.github.io/IntersectionObserver/#notify-intersection-observers-algo>
+    pub(crate) fn invoke_callback(&self) {
+        // Step 1
+        // > If observer’s internal [[QueuedEntries]] slot is empty, continue.
+        if self.queued_entries.borrow().is_empty() {
+            return;
+        }
+
+        // Step 2-3
+        // We trivially moved the entries and root them.
+        let queued_entries = self
+            .queued_entries
+            .take()
+            .iter_mut()
+            .map(|entry| entry.as_rooted())
+            .collect();
+
+        // Step 4-5
+        let _ = self
+            .callback
+            .Call_(self, queued_entries, self, ExceptionHandling::Report);
+    }
+
+    /// Connect the observer itself into owner doc if it is unconnected.
+    fn connect_to_owner_unchecked(&self) {
+        if self.is_connected.get() {
+            return;
+        }
+        self.owner_doc.add_intersection_observer(self);
+        self.is_connected.set(true);
+    }
+
+    /// Disconnect the observer itself from owner doc.
+    fn disconnect_from_owner_unchecked(&self) {
+        if !self.is_connected.get() {
+            return;
+        }
+        self.owner_doc.remove_intersection_observer(self);
+        self.is_connected.set(false);
+    }
+
+    /// Disconnect the observer itself from owner doc. But check whether target exist or not.
+    fn disconnect_from_owner(&self) {
+        if !self.observation_targets.borrow().is_empty() {
+            self.disconnect_from_owner_unchecked();
+        }
+    }
+
+    /// > The root intersection rectangle for an IntersectionObserver is
+    /// > the rectangle we’ll use to check against the targets.
+    ///
+    /// NOTE: Firefox seems to only account for viewport/boundingBox area without not covered by
```

Comment:
It would be useful to point out why FF's behavior is relevant here. I assume because we do the same?

Reply:
Currently, it is using the `viewport` that is quite different from the spec and other implementation. It is considering the scroll offset and not considering the size of the scrollbar. I really should add this into another todo, individually on each impacted lines.

Reply:
I have checked this further, it seems that FF behavior make more sense. And, the spec was quite incomplete in this section. I believe it should consider clipping and probably several more steps similar to https://w3c.github.io/IntersectionObserver/#compute-the-intersection. I will check chromium's implementation too, as FF is not quite updated regarding several new features. 

I will leave viewport addressing similar to element [client-width](https://drafts.csswg.org/cssom-view/#dom-element-clientwidth) for now.

Reply:
Ok, thanks for all the research. For clarity, what I mean is that we should write the comment from the perspective of what Servo is doing, and then perhaps explain that this matches FF's behavior and that the spec has not settled down yet. SO basically the same content, but putting the focus on what we're trying to do.

Reply:
Thanks for pointing out. Sorry for the misunderstanding the intention.